File: Java/maximum-subarray.java
//  Maximum Subarray

// Time O(N), Space O(1)
class Solution {
    public int maxSubArray(int[] nums) {
        int max_sum = Integer.MIN_VALUE, sub_sum = 0;
        for (int i=0; i<nums.length; i++) {
            sub_sum = Math.max(sub_sum+nums[i], nums[i]);
            max_sum = Math.max(max_sum, sub_sum);
        }
        return max_sum;
    }
}

File: Leetcode150/maximum-subarray.cc
//  Maximum Subarray

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int res = INT_MIN, sub_sum = 0;
        for (int &x : nums) {
            sub_sum = max(sub_sum + x, x);
            res = max(sub_sum, res);
        }
        return res;
    }
};